Source/WebCore/css/CSSValuePair.h

```diff
@@ -25,39 +26,64 @@
 
 #pragma once
 
-#include "CSSValue.h"
+#include "CSSIdentValue.h"
 
 namespace WebCore {
 
 class CSSValuePair : public CSSValue {
 public:
-    static Ref<CSSValuePair> create(Ref<CSSValue>, Ref<CSSValue>);
-    static Ref<CSSValuePair> createNoncoalescing(Ref<CSSValue>, Ref<CSSValue>);
+    static Ref<CSSValuePair> create(Ref<CSSValue>&&, Ref<CSSValue>&&);
+    static Ref<CSSValuePair> createNoncoalescing(Ref<CSSValue>&&, Ref<CSSValue>&&);
+    static Ref<CSSValuePair> create(CSSValueID, CSSValueID);
 
-    const CSSValue& first() const { return m_first; }
-    const CSSValue& second() const { return m_second; }
+    const CSSValue& first() const;
+    const CSSValue& second() const;
+    bool isNoncoalescing() const;
 
     String customCSSText() const;
     bool equals(const CSSValuePair&) const;
 
 private:
+    // Encode values in the scalar.
+    static constexpr uintptr_t IsNoncoalescingBit = 1 << PayloadShift;
+    static constexpr uint8_t FirstValueIDShift = PayloadShift + 1;
+    static constexpr uint8_t SecondValueIDShift = FirstValueIDShift + IdentBits;
+
     enum class IdenticalValueSerialization : bool { DoNotCoalesce, Coalesce };
-    CSSValuePair(Ref<CSSValue>, Ref<CSSValue>, IdenticalValueSerialization);
+    CSSValuePair(Ref<CSSValue>&&, Ref<CSSValue>&&, IdenticalValueSerialization);
 
-    // FIXME: Store coalesce bit in CSSValue to cut down on object size.
-    bool m_coalesceIdenticalValues { true };
     Ref<CSSValue> m_first;
     Ref<CSSValue> m_second;
 };
 
-inline const CSSValue& CSSValue::first() const
+inline const CSSValue& CSSValuePair::first() const
+{
+    if (hasScalarInPointer())
+        return CSSIdentValue::create(static_cast<CSSValueID>(scalar() >> FirstValueIDShift & IdentMask)).get();
+    return opaque(this)->m_first.get();
+}
```

Comment:
Regarding `opaque()`, it’s needed for two reasons:

1) Anywhere we actually use the pointer, we must call `opaque()` so the compiler can’t make any assumptions about the pointer’s value. That’s because otherwise it can assume the pointer is non-null and dereference it, which is bad since it’s not a pointer. Using `opaque()` means it doesn’t make such assumptions. It can assume a this is non-null and it can assume a pointer that came from a reference is non-null. So in both cases, the actual use of the pointer has to be after a call to `opaque()`.

2) Anywhere we check the low bit, we must call `opaque()` to hide that the scalar came from a pointer. That’s because otherwise the compiler can assume pointer alignment can optimize out the branch, assuming it is zero.

I should write a comment explaining this somewhere in the code.